### PR TITLE
fix: populate context from config

### LINKED ISSUE #451: ctx.data and ctx.assets are not populated by the config
In the `beet_default` of a custom plugin, the config values for `data_pack` and `resource_pack` are not inserted into the `ctx.data` and `ctx.assets` fields respectively. The config values are currently not available except for the internal plugin that generates the output (i.e. the `pack.mcmeta` file).

### Example:
`beet.yaml`
```yaml
data_pack:
  name: My Data Pack
  description: a data pack
  load: src
  pack_format: 6
  supported_formats: [0,6]

pipeline:
  - custom_plugin
```

`custom_plugin.py`
```py
def beet_default(ctx: Context):
  print(ctx.data.name)
  print(ctx.data.description)
  print(ctx.data.pack_format)
  print(ctx.data.supported_formats)
```

**Result**
This will print out the following (i.e. the default values):
```
None

48
None
```

**Expected**
The values should return the value set in the config:
```bash
My Data Pack
a data pack
6
[0, 6]
```

### PR DESCRIPTION
**Pull request by BPR02**
- fix #451 
- populate ctx.data and ctx.assets with the following info from the config:
  - name
  - description
  - pack_format
  - supported_formats

### CODE REVIEW

**Comment by BPR02** [comment]
@misode @RitikShah @edayot I think this PR can be merged. I tested it with my Observer plugin and it seems to be working as expected. I'd like a review if you guys have time.

I added some tests, but there might be a more elegant way to test the output... I'm not super familiar with pytest, but I couldn't find anything online about passing an output to an object for pytest to catch, so I just put the text into a meta field and checked it afterwards.

**Review by edayot — CHANGES_REQUESTED** [review]

File: tests/test_ctx_config.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import pytest
+from pytest_insta import SnapshotFixture
+
+from beet import run_beet
+
+
+@pytest.mark.parametrize("directory", os.listdir("tests/ctx_config_examples"))
+def test_examples(snapshot: SnapshotFixture, directory: str):
+    with run_beet(directory=f"tests/ctx_config_examples/{directory}") as ctx:
+        assert snapshot() == f'{ctx.meta["pytest"]}'
```
[the remark this answers is not in the document]
Comment (edayot) [reply tests/test_ctx_config.py:12]: Oh i see, i don't like the `ctx.meta["pytest"]` access, but it's the only way i can think of accessing plugin phase value.

**Review by edayot — APPROVED** [review]

File: tests/test_ctx_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BPR02) [reply tests/test_ctx_config.py:12]: Yeah, that was the main reason I wanted a review; the `ctx.meta["pytest"]` feels really janky, but I couldn't find another way to check those values.

**Comment by misode** [comment]
@BPR02 I have made a small adjustment: I've moved this copying logic to the `ProjectBuilder.bootstrap` function, since that is also where the other `data_pack` and `resource_pack` loading and outputting is done. To me it also makes sense that this data would only be used after `require`.

Since the order is currently (1) `require` -> (2) `data|resource_pack load` -> (3) `data|resource_pack render` -> (4) `pipeline` -> (5) `data|resource_pack output`. What this PR now does is insert a set between (1) and (2) that copies the `data|resource_pack` config to the `ctx`. It could've also been between (2) and (3) but I don't think it makes any difference. It being before (3) is probably useful if you want to use these variables in a jinja template.

### Further remarks

1. We can probably also do this for the part that sets the `description_parts` (lines 340 to 359, and line 361), which would allow plugins to overwrite the pack description. I vaguely remember encountering such a problem for GM4. Should probably not be in this PR though.
2. At line 373 and below there is already logic for populating the pack's data from the config. Maybe we can just move part of this logic to the top instead of doing it twice?

**Comment by misode** [comment]
I've made some more drastic changes now, and it's not quite done yet, because there is some behavior change, for example when pack metadata is present both in the beet config and in `pack.mcmeta`.

- Having the `pack.pack_format` be set to the beet config `pack_format` is useful when loading from a folder that doesn't a `pack.mcmeta` but you still want the right scopes to be used. I am not sure how beet currently handles this (needs to be tested)
- On the other hand, this means that if `pack_format` is defined in both the beet config and `pack.mcmeta`, with this PR `pack.mcmeta` will take priority. Similarly for both `filter` and `overlays`, the entries from `pack.mcmeta` will be appended after the values from the beet config.

With regards to which config takes priority, I am unsure whether we should keep compatibility with the released beet version, or decide from scratch what the ideal behavior should be.

Todo: add tests for all these edge cases

**Comment by BPR02** [comment]
I think it makes sense to make the beet config be the "one true source" of configuration values. I'm not sure what the current release behavior is, but for this PR maybe it should be changed to only pull the `pack_format` from the pack.mcmeta if it's not already defined in the beet config. However I think `filter` and `overlays` would make sense to append. Would it be confusing to users if some values are overwritten (e.g. `pack_format`) while others are extended (e.g. `overlays`)?

**Comment by rx-dev** [comment]
@misode @BPR02 doing cleanup on prs, do u think we can get this PR fixed up and merged?
